Add remote caching option for build.ps1

brianmichel opened the pull request:
When setting up sccache we can choose between local disk caching or remote caching via an S3-ish blob storage engine. To allow for backward compatibility for anyone that might be using the existing option for disk caching we leave it place and instead provide a new remote caching option.

This option assumes you've read the sccache documentation for setting up the correct environment variables for telling sccache where to find the bucket and how to authenticate to it.

brianmichel committed:
Add remote caching option

When setting up sccache we can choose between local disk caching or
remote caching via an S3-ish blob storage engine. To allow for backward
compatibility for anyone that might be using the existing option for
disk caching we leave it place and instead provide a new remote caching
option.

This option assumes you've read the sccache documentation for setting up
the correct environment variables for telling sccache where to find the
bucket and how to authenticate to it.

brianmichel commented:
cc @compnerd

brianmichel commented on utils/build.ps1:
```diff
@@ -167,9 +171,13 @@ param
   [string] $Variant = "Asserts",
   [switch] $Clean,
   [switch] $DebugInfo,
+  [parameter(ParameterSetName = "LocalCaching")]
```
Using parameter set names here to isolate the local options from the remote options.

compnerd reviewed (changes requested):
I'd rather not use `ParameterSet` here. You can only have a single parameter set, and I would rather that if we were to use it, it would apply to the pinned toolchain parameters.

What is the advantage to having `-EnableRemoteCaching`? Why not have a single `-EnableCaching` and make the user responsible for configuring the cache?

brianmichel commented:
> I'd rather not use `ParameterSet` here. You can only have a single parameter set, and I would rather that if we were to use it, it would apply to the pinned toolchain parameters.
> 
> What is the advantage to having `-EnableRemoteCaching`? Why not have a single `-EnableCaching` and make the user responsible for configuring the cache?

@compnerd Could certainly do that! @jeffdav and I were talking and thought that maybe people are using `-EnableCaching` downstream and assume a specific behavior. I'm happy to rewrite those expectations. IMO `-EnableCaching` probably should have ever only just setup the sccache launchers in CMake. We defaulted to backward compatibility because that is the Windows way, lmk if you think it's okay to just reset expectations of `EnableCaching`!

compnerd commented:
The expected behavior AFAIK is that `-EnableCaching` will use `sccache`. The configuration of sccache itself however is beyond the scope of `build.ps1`.

brianmichel committed:
Remove cache configuration from EanbleCaching

EnableCaching should only configure cmake to allow for sccache usage, it
should not configure sccache itself in any way. Those configuration
options should be deferred into the environment variables that sccache
will look for when it starts up.

compnerd reviewed (approved):
I think that this makes sense.
